### PR TITLE
The number of threads for loading and unloading models is set to 1

### PR DESCRIPTION
**Pull request by handong0123**
`uint32 num_load_threads = 0;`
`uint32 num_unload_threads = 0;`

These two parameters are 0 by default, which means that independent thread pools are not used and run in the same thread as Serving Manager.

The threads that update, load the model, and process TensorFlow Serving requests share the same thread pool, which makes it impossible to process requests when switching models.

### CODE REVIEW

**Comment by netfs** [comment]
> The threads that update, load the model, and process TensorFlow Serving requests share the same thread pool, which makes it impossible to process requests when switching models.

i can see the value of using different threadpools for isolation, but cannot see why it becomes "impossible" to process requests while model is loading (if that is what you mean). i'd expect >1 threads in the common pool, that will allow to serve requests and load model, yes at times causing serving requests to starve or slow down.

but "impossible" sounds a bit odd. care to explain more?

thanks!

**Comment by handong0123** [comment]
> > The threads that update, load the model, and process TensorFlow Serving requests share the same thread pool, which makes it impossible to process requests when switching models.
> 
> i can see the value of using different threadpools for isolation, but cannot see why it becomes "impossible" to process requests while model is loading (if that is what you mean). i'd expect >1 threads in the common pool, that will allow to serve requests and load model, yes at times causing serving requests to starve or slow down.
> 
> but "impossible" sounds a bit odd. care to explain more?
> 
> thanks!

Maybe my description is inaccurate, The core operation of model loading is RestoreOp, which includes operations such as reading model files from storage, allocating memory, and finding the corresponding Variable, which are executed by calling the run method of Session. By default, all Session operations in a process use the same thread pool. Therefore, the loading operation and the operation processing the Serving request in the model loading process use the same thread pool, which causes the Serving request to be delayed.So there will be glitches when the model is switched.

**Comment by netfs** [comment]
> > > The threads that update, load the model, and process TensorFlow Serving requests share the same thread pool, which makes it impossible to process requests when switching models.
> > 
> > 
> > i can see the value of using different threadpools for isolation, but cannot see why it becomes "impossible" to process requests while model is loading (if that is what you mean). i'd expect >1 threads in the common pool, that will allow to serve requests and load model, yes at times causing serving requests to starve or slow down.
> > but "impossible" sounds a bit odd. care to explain more?
> > thanks!
> 
> Maybe my description is inaccurate, The core operation of model loading is RestoreOp, which includes operations such as reading model files from storage, allocating memory, and finding the corresponding Variable, which are executed by calling the run method of Session. By default, all Session operations in a process use the same thread pool. Therefore, the loading operation and the operation processing the Serving request in the model loading process use the same thread pool, which causes the Serving request to be delayed.So there will be glitches when the model is switched.

Thanks. That helped clarify. Yes things can get delayed or starve due to having one threadpool for loading+serving.

File: tensorflow_serving/core/aspired_versions_manager.h [hunk 1]
```diff
@@ -110,13 +110,13 @@ class AspiredVersionsManager : public Manager,
     ///
     /// If set as 0, we don't use a thread-pool, and servable loads are
     /// performed serially in the manager's main work loop.
-    uint32 num_load_threads = 0;
+    uint32 num_load_threads = 1;
```
Comment (netfs) [inline tensorflow_serving/core/aspired_versions_manager.h:114]: please do not change default values. instead add two new flags to the server (`num_load_threads` and `num_unload_threads`), so people can (optionally) tweak the settings as needed for their use case.

as an example, copy/follow the pattern on how `--max_num_load_retries` flag is implemented/used to set appropriate lower level options.

https://github.com/tensorflow/serving/blob/0acaa63b45a5c461dcb4808f8745ff9088a0f6e6/tensorflow_serving/model_servers/main.cc#L108

you can add new flags around this.

Comment (handong0123) [reply tensorflow_serving/core/aspired_versions_manager.h]: ok, I had added two new flags to the server (num_load_threads and num_unload_threads).

File: tensorflow_serving/core/basic_manager.cc [hunk 2]
```diff
@@ -230,8 +230,8 @@ Status BasicManager::Create(Options options,
   return Status::OK();
 }
 
-BasicManager::BasicManager(Env* const env, const uint32 num_load_threads,
```
Comment (netfs) [inline tensorflow_serving/core/basic_manager.cc:233]: why do we need this change?

Comment (handong0123) [reply tensorflow_serving/core/basic_manager.cc:233]: I don’t quite understand why max_num_load_retries is not modified by const, so in order to be consistent, I made changes, which may not be appropriate.

Comment (netfs) [reply tensorflow_serving/core/basic_manager.cc:233]: i'd recommend reverting this file, as i dont see it related to the change you are doing. can you please revert this file?

also can you please update the commit message to read better -- mention the new flags that you have added, and what they do.

thanks!

Comment (handong0123) [reply tensorflow_serving/core/basic_manager.cc]: ok, I've completed.

**Comment by netfs** [comment]
thanks for the change and your contribution. will approve and merge shortly.

File: tensorflow_serving/model_servers/main.cc [hunk 3]
```diff
@@ -105,6 +105,18 @@ int main(int argc, char** argv) {
       tensorflow::Flag("model_base_path", &options.model_base_path,
                        "path to export (ignored if --model_config_file flag "
                        "is set, otherwise required)"),
+      tensorflow::Flag("num_load_threads", &options.num_load_threads,
+                       "The number of threads in the thread-pool used to load servables."
+                       "If set as 0, we don't use a thread-pool, and servable loads are "
+                       "performed serially in the manager's main work loop, "
+                       "may casue the Serving request to be delayed. "
+                       "Default: 0"),`
```
Comment (netfs) [inline tensorflow_serving/model_servers/main.cc:113]: there is a stray ``` at the end. please remove. this is i think causing the tests to break:

https://source.cloud.google.com/results/invocations/dc6486cd-9c79-486b-b5db-0632da419481/targets

can you also ensure all the tests pass?

thanks!

Comment (handong0123) [reply tensorflow_serving/model_servers/main.cc]: This is my mistake, now corrected

**Review by netfs — APPROVED** [review]

**Comment by netfs** [comment]
Thanks!

**Pull request merged**